Docker Slaves die randomly during build.

**Docker Plugin versions:**
    1.1.3 *and* 1.2 SNAPSHOT
	
**Jenkins Version:** 
    2.89.4
	
**Docker Version:**
    Version:      17.03.2-ce
    API version:  1.27 (minimum version 1.12)
    Go version:   go1.7.5
    Git commit:   f5ec1e2
    Built:        Tue Jun 27 02:21:36 2017

**Possibly related issue:**
    https://github.com/jenkinsci/docker-plugin/issues/621

**My actual problem**

I created a basic freestyle test project. This just prints out a directory listing, touches a file (to make sure mounting volumes works correctly), but doesn't run for >60 seconds. 
This always completes correctly.

With all versions of the plugin, what happens, other than for the trivial "Test" job, all jobs fail before completion.

All failing jobs are Maven based.

- Some can run for 30 seconds then fail while cloning the git repository.
- Some can die after 5 minutes. The last stack trace below shows this. I'm only building a single bundle, as it's gets frustrating when the build fails 2 minutes before the end of a 50 minute release, so by using just one bundle, it fails more quickly.
- Some can run for 45 minutes, and then fail when Jenkins is cleaning up. The same job which gave the last stack trace normally fails after 10 minutes (give or take)

When they die, they are always cleaned up correctly, and the next job always starts correctly.

I upgraded to 1.2 SNAPSHOT after I saw the possibly related issue above, but it made no difference.

Infrastructure is a single VM, dual core 8Gb RAM, 100Gb HDD running the docker server + JFrog artifactory.

There is no old data present currently. I had an issue with old maven data being present, but that's been removed.

**Relevant Jenkins Config**

Template is configured to communicate via Attaching to Docker Container. This is because if I connect via SSH, something, somewhere has an issue, and the SSH key is passed from the Docker Plugin>Docker Container to the Container>GIT repository, which causes an authentication failure. I can't build a JNLP image on our server here, due to missing files, so I'm kind of stuck with the SSH image for now.

Configuration for the Jenkins Template is below. idleMinutes is set to 10, but it seems to make no difference if I set it to 60.

Clouds configuration is attached as clouds.txt.

**Installation History**
Version 1.1.2 was already installed on the box when I took over managing it. I then upgraded to 1.1.3. **As far as I know**, version 0.16.x or earlier has never been installed on this Jenkins.

I've installed 1.2 SNAPSHOT to see if it helped.


**Stack Traces:**
All stack traces are along the lines of the blocks below. I've not supplied the full build logs, as most of these are >1Mb, and don't actually give much information.

> 14:00:49 Archiving artifacts
> 14:00:52 ERROR: Failed to archive artifacts: Installer/MiddlewareInstallerKarafAssembly/*.exe, Installer/MiddlewareInstallerKarafAssembly/*.run, Installer/MiddlewareInstallerKarafAssembly/AutoInstallScript.txt, Middleware-Parent/karaf-assembly/license.txt
> 14:00:52 java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 4
> 14:00:52 	at io.jenkins.docker.client.DockerMultiplexedInputStream.readInternal(DockerMultiplexedInputStream.java:45)
> 14:00:52 	at io.jenkins.docker.client.DockerMultiplexedInputStream.read(DockerMultiplexedInputStream.java:25)
> 14:00:52 	at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:91)
> 14:00:52 	at hudson.remoting.ChunkedInputStream.readHeader(ChunkedInputStream.java:72)
> 14:00:52 	at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:103)
> 14:00:52 	at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:39)
> 14:00:52 	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)
> 14:00:52 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)
> 14:00:52 Caused: java.io.IOException
> 14:00:52 	at hudson.remoting.FastPipedInputStream.read(FastPipedInputStream.java:169)
> 14:00:52 	at hudson.util.HeadBufferingStream.read(HeadBufferingStream.java:61)
> 14:00:52 	at com.jcraft.jzlib.InflaterInputStream.fill(InflaterInputStream.java:175)
> 14:00:52 	at com.jcraft.jzlib.InflaterInputStream.read(InflaterInputStream.java:106)
> 14:00:52 	at org.apache.commons.compress.archivers.tar.TarArchiveInputStream.read(TarArchiveInputStream.java:614)
> 14:00:52 	at java.io.InputStream.read(Unknown Source)
> 14:00:52 	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1792)
> 14:00:52 	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1769)
> 14:00:52 	at org.apache.commons.io.IOUtils.copy(IOUtils.java:1744)
> 14:00:52 	at hudson.util.IOUtils.copy(IOUtils.java:40)
> 14:00:52 	at hudson.FilePath.readFromTar(FilePath.java:2307)
> 14:00:52 Also:   <cycle to java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 4>
> 14:00:52 	Caused: hudson.remoting.RequestAbortedException
> 14:00:52 		at hudson.remoting.Request.abort(Request.java:329)
> 14:00:52 		at hudson.remoting.Channel.terminate(Channel.java:992)
> 14:00:52 		at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:96)
> 14:00:52 	Caused: java.util.concurrent.ExecutionException
> 14:00:52 		at hudson.remoting.Request$1.get(Request.java:308)
> 14:00:52 		at hudson.remoting.Request$1.get(Request.java:229)
> 14:00:52 		at hudson.remoting.FutureAdapter.get(FutureAdapter.java:59)
> 14:00:52 		at hudson.FilePath.copyRecursiveTo(FilePath.java:2224)
> 14:00:52 Caused: java.io.IOException: Failed to extract /home/jenkins/workspace/Middleware_Develop/transfer of 4 files
> 14:00:52 	at hudson.FilePath.readFromTar(FilePath.java:2317)
> 14:00:52 	at hudson.FilePath.copyRecursiveTo(FilePath.java:2221)
> 14:00:52 	at jenkins.model.StandardArtifactManager.archive(StandardArtifactManager.java:61)
> 14:00:52 	at hudson.tasks.ArtifactArchiver.perform(ArtifactArchiver.java:245)
> 14:00:52 	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:81)
> 14:00:52 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
> 14:00:52 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:744)
> 14:00:52 	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:690)
> 14:00:52 	at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.post2(MavenModuleSetBuild.java:1073)
> 14:00:52 	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:635)
> 14:00:52 	at hudson.model.Run.execute(Run.java:1749)
> 14:00:52 	at hudson.maven.MavenModuleSetBuild.run(MavenModuleSetBuild.java:543)
> 14:00:52 	at hudson.model.ResourceController.execute(ResourceController.java:97)
> 14:00:52 	at hudson.model.Executor.run(Executor.java:429)
> 14:00:52 Build step 'Archive the artifacts' changed build result to FAILURE



> 14:19:47 ERROR: Build step failed with exception
> 14:19:47 java.lang.NullPointerException: no workspace from node io.jenkins.docker.DockerTransientNode[docker-298a19d5038f4] which is computer DockerComputer{name=docker-298a19d5038f4, slave=io.jenkins.docker.DockerTransientNode[docker-298a19d5038f4]} and has channel null
> 14:19:47 	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:88)
> 14:19:47 	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:66)
> 14:19:47 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
> 14:19:47 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:744)
> 14:19:47 	at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.build(MavenModuleSetBuild.java:945)
> 14:19:47 	at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.doRun(MavenModuleSetBuild.java:896)
> 14:19:47 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:504)
> 14:19:47 	at hudson.model.Run.execute(Run.java:1724)
> 14:19:47 	at hudson.maven.MavenModuleSetBuild.run(MavenModuleSetBuild.java:543)
> 14:19:47 	at hudson.model.ResourceController.execute(ResourceController.java:97)
> 14:19:47 	at hudson.model.Executor.run(Executor.java:429)
> 14:19:47 Build step 'Execute shell' marked build as failure
> 14:19:47 ERROR: Failed to parse POMs
> 14:19:47 java.io.IOException: Failed to extract /home/jenkins/workspace/Middleware_Develop/Middleware-Parent/karaf-assembly/transfer of 3 files
> 14:19:47 	at hudson.FilePath.readFromTar(FilePath.java:2317)
> 14:19:47 	at hudson.FilePath.copyRecursiveTo(FilePath.java:2221)
> 14:19:47 	at jenkins.model.StandardArtifactManager.archive(StandardArtifactManager.java:61)
> 14:19:47 	at hudson.maven.MavenBuild$ProxyImpl.performArchiving(MavenBuild.java:512)
> 14:19:47 	at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.doRun(MavenModuleSetBuild.java:881)
> 14:19:47 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:504)
> 14:19:47 	at hudson.model.Run.execute(Run.java:1724)
> 14:19:47 	at hudson.maven.MavenModuleSetBuild.run(MavenModuleSetBuild.java:543)
> 14:19:47 	at hudson.model.ResourceController.execute(ResourceController.java:97)
> 14:19:47 	at hudson.model.Executor.run(Executor.java:429)
> 14:19:47 	Suppressed: java.util.concurrent.ExecutionException: hudson.remoting.RequestAbortedException: java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 7
> 14:19:47 		at hudson.remoting.Request$1.get(Request.java:308)
> 14:19:47 		at hudson.remoting.Request$1.get(Request.java:229)
> 14:19:47 		at hudson.remoting.FutureAdapter.get(FutureAdapter.java:59)
> 14:19:47 		at hudson.FilePath.copyRecursiveTo(FilePath.java:2224)
> 14:19:47 		... 8 more
> 14:19:47 	Caused by: hudson.remoting.RequestAbortedException: java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 7
> 14:19:47 		at hudson.remoting.Request.abort(Request.java:329)
> 14:19:47 		at hudson.remoting.Channel.terminate(Channel.java:992)
> 14:19:47 		at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:96)
> 14:19:47 	Caused by: java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 7
> 14:19:47 		at io.jenkins.docker.client.DockerMultiplexedInputStream.readInternal(DockerMultiplexedInputStream.java:45)
> 14:19:47 		at io.jenkins.docker.client.DockerMultiplexedInputStream.read(DockerMultiplexedInputStream.java:25)
> 14:19:47 		at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:91)
> 14:19:47 		at hudson.remoting.ChunkedInputStream.readHeader(ChunkedInputStream.java:72)
> 14:19:47 		at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:103)
> 14:19:47 		at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:39)
> 14:19:47 		at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)
> 14:19:47 		at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)
> 14:19:47 Caused by: java.io.IOException: java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 7
> 14:19:47 	at hudson.remoting.FastPipedInputStream.read(FastPipedInputStream.java:169)
> 14:19:47 	at hudson.util.HeadBufferingStream.read(HeadBufferingStream.java:61)
> 14:19:47 	at com.jcraft.jzlib.InflaterInputStream.fill(InflaterInputStream.java:175)
> 14:19:47 	at com.jcraft.jzlib.InflaterInputStream.read(InflaterInputStream.java:106)
> 14:19:47 	at org.apache.commons.compress.archivers.tar.TarArchiveInputStream.read(TarArchiveInputStream.java:614)
> 14:19:47 	at java.io.InputStream.read(Unknown Source)
> 14:19:47 	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1792)
> 14:19:47 	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1769)
> 14:19:47 	at org.apache.commons.io.IOUtils.copy(IOUtils.java:1744)
> 14:19:47 	at hudson.util.IOUtils.copy(IOUtils.java:40)
> 14:19:47 	at hudson.FilePath.readFromTar(FilePath.java:2307)
> 14:19:47 	... 9 more
> 14:19:47 	[CIRCULAR REFERENCE:java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 7]
> 14:19:48 ERROR: Step ‘JIRA: Update relevant issues’ failed: no workspace for Middleware_Develop #42



> 20:20:36 Waiting for Jenkins to finish collecting data
> 20:20:36 [JENKINS] Archiving /home/jenkins/workspace/Quareo-Simulator_QS-Release/qrc-firmware-manager/pom.xml to com.te.quareosimulator.firmware/qrc-firmware-manager/2.2.5-RELEASE/qrc-firmware-manager-2.2.5-RELEASE.pom
> 20:20:36 [JENKINS] Archiving /home/jenkins/workspace/Quareo-Simulator_QS-Release/qrc-firmware-manager/target/qrc-firmware-manager.jar to com.te.quareosimulator.firmware/qrc-firmware-manager/2.2.5-RELEASE/qrc-firmware-manager-2.2.5-RELEASE.jar
> 20:21:01 [JENKINS] Archiving /home/jenkins/workspace/Quareo-Simulator_QS-Release/config-rest-interface/pom.xml to com.te.quareosimulator.endpoint/config-rest-endpoints/2.2.5-RELEASE/config-rest-endpoints-2.2.5-RELEASE.pom
> 20:21:01 [JENKINS] Archiving /home/jenkins/workspace/Quareo-Simulator_QS-Release/config-rest-interface/target/config-rest-endpoints.jar to com.te.quareosimulator.endpoint/config-rest-endpoints/2.2.5-RELEASE/config-rest-endpoints-2.2.5-RELEASE.jar
> 20:21:02 [JENKINS] Archiving /home/jenkins/workspace/Quareo-Simulator_QS-Release/quareo-device-cache/pom.xml to com.te.quareosimulator.device/quareo-device-cache/2.2.5-RELEASE/quareo-device-cache-2.2.5-RELEASE.pom
> 20:21:02 [JENKINS] Archiving /home/jenkins/workspace/Quareo-Simulator_QS-Release/quareo-device-cache/target/quareo-device-cache.jar to com.te.quareosimulator.device/quareo-device-cache/2.2.5-RELEASE/quareo-device-cache-2.2.5-RELEASE.jar
> 20:21:03 [JENKINS] Archiving /home/jenkins/workspace/Quareo-Simulator_QS-Release/karaf-assembly/pom.xml to com.te.quareosimulator.karaf/karaf-assembly/2.2.5-RELEASE/karaf-assembly-2.2.5-RELEASE.pom
> 20:21:03 [JENKINS] Archiving /home/jenkins/workspace/Quareo-Simulator_QS-Release/karaf-assembly/pom.xml to com.te.quareosimulator.karaf/karaf-assembly/2.2.5-RELEASE/karaf-assembly-2.2.5-RELEASE.karaf-assembly
> 20:21:03 [JENKINS] Archiving /home/jenkins/workspace/Quareo-Simulator_QS-Release/karaf-assembly/target/karaf-assembly-2.2.5-RELEASE.tar.gz to com.te.quareosimulator.karaf/karaf-assembly/2.2.5-RELEASE/karaf-assembly-2.2.5-RELEASE.tar.gz
> 20:21:03 [JENKINS] Archiving /home/jenkins/workspace/Quareo-Simulator_QS-Release/karaf-assembly/target/karaf-assembly-2.2.5-RELEASE.zip to com.te.quareosimulator.karaf/karaf-assembly/2.2.5-RELEASE/karaf-assembly-2.2.5-RELEASE.zip
> 20:21:05 ERROR: Failed to parse POMs
> 20:21:05 java.io.IOException: Failed to extract /home/jenkins/workspace/Quareo-Simulator_QS-Release/karaf-assembly/transfer of 4 files
> 20:21:05 	at hudson.FilePath.readFromTar(FilePath.java:2317)
> 20:21:05 	at hudson.FilePath.copyRecursiveTo(FilePath.java:2221)
> 20:21:05 	at jenkins.model.StandardArtifactManager.archive(StandardArtifactManager.java:61)
> 20:21:05 	at hudson.maven.MavenBuild$ProxyImpl.performArchiving(MavenBuild.java:512)
> 20:21:05 	at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.doRun(MavenModuleSetBuild.java:881)
> 20:21:05 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:504)
> 20:21:05 	at hudson.model.Run.execute(Run.java:1724)
> 20:21:05 	at hudson.maven.MavenModuleSetBuild.run(MavenModuleSetBuild.java:543)
> 20:21:05 	at hudson.model.ResourceController.execute(ResourceController.java:97)
> 20:21:05 	at hudson.model.Executor.run(Executor.java:429)
> 20:21:05 	Suppressed: java.util.concurrent.ExecutionException: hudson.remoting.RequestAbortedException: java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 1
> 20:21:05 		at hudson.remoting.Request$1.get(Request.java:308)
> 20:21:05 		at hudson.remoting.Request$1.get(Request.java:229)
> 20:21:05 		at hudson.remoting.FutureAdapter.get(FutureAdapter.java:59)
> 20:21:05 		at hudson.FilePath.copyRecursiveTo(FilePath.java:2224)
> 20:21:05 		... 8 more
> 20:21:05 	Caused by: hudson.remoting.RequestAbortedException: java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 1
> 20:21:05 		at hudson.remoting.Request.abort(Request.java:329)
> 20:21:05 		at hudson.remoting.Channel.terminate(Channel.java:992)
> 20:21:05 		at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:96)
> 20:21:05 	Caused by: java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 1
> 20:21:05 		at io.jenkins.docker.client.DockerMultiplexedInputStream.readInternal(DockerMultiplexedInputStream.java:45)
> 20:21:05 		at io.jenkins.docker.client.DockerMultiplexedInputStream.read(DockerMultiplexedInputStream.java:25)
> 20:21:05 		at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:91)
> 20:21:05 		at hudson.remoting.ChunkedInputStream.readHeader(ChunkedInputStream.java:72)
> 20:21:05 		at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:103)
> 20:21:05 		at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:39)
> 20:21:05 		at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)
> 20:21:05 		at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)
> 20:21:05 Caused by: java.io.IOException: java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 1
> 20:21:05 	at hudson.remoting.FastPipedInputStream.read(FastPipedInputStream.java:169)
> 20:21:05 	at hudson.util.HeadBufferingStream.read(HeadBufferingStream.java:61)
> 20:21:05 	at com.jcraft.jzlib.InflaterInputStream.fill(InflaterInputStream.java:175)
> 20:21:05 	at com.jcraft.jzlib.InflaterInputStream.read(InflaterInputStream.java:106)
> 20:21:05 	at org.apache.commons.compress.archivers.tar.TarArchiveInputStream.read(TarArchiveInputStream.java:614)
> 20:21:05 	at java.io.InputStream.read(Unknown Source)
> 20:21:05 	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1792)
> 20:21:05 	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1769)
> 20:21:05 	at org.apache.commons.io.IOUtils.copy(IOUtils.java:1744)
> 20:21:05 	at hudson.util.IOUtils.copy(IOUtils.java:40)
> 20:21:05 	at hudson.FilePath.readFromTar(FilePath.java:2307)
> 20:21:05 	... 9 more
> 20:21:05 	[CIRCULAR REFERENCE:java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 1]
> 20:21:05 ERROR: Step ‘Archive the artifacts’ failed: no workspace for Quareo-Simulator_QS-Release #67

This stack trace shows it happening 5 minutes after the start of the build...

> Started by user Smith, Peter (Pete)
> [EnvInject] - Loading node environment variables.
> Building remotely on docker-29a83953db3cf on docker (docker-Middleware_Develop) in workspace /home/jenkins/workspace/Middleware_Develop
>  > /usr/bin/git rev-parse --is-inside-work-tree # timeout=10
> Fetching changes from the remote Git repository
>  > /usr/bin/git config remote.origin.url http://cdcaimgitp1.commscope.com/root/QuareoMiddleware.git # timeout=10
> Fetching upstream changes from http://cdcaimgitp1.commscope.com/root/QuareoMiddleware.git
>  > /usr/bin/git --version # timeout=10
> using GIT_ASKPASS to set credentials Gitlab Checkout
>  > /usr/bin/git fetch --tags --progress http://cdcaimgitp1.commscope.com/root/QuareoMiddleware.git +refs/heads/*:refs/remotes/origin/* # timeout=30
>  > /usr/bin/git rev-parse refs/remotes/origin/develop^{commit} # timeout=10
>  > /usr/bin/git rev-parse refs/remotes/origin/origin/develop^{commit} # timeout=10
> Checking out Revision 5f95be55e3327417cea16f2cf418bad7fa91b705 (refs/remotes/origin/develop)
>  > /usr/bin/git config core.sparsecheckout # timeout=10
>  > /usr/bin/git checkout -f 5f95be55e3327417cea16f2cf418bad7fa91b705
> Commit message: "AIMQS-155 Further changes to config files"
>  > /usr/bin/git rev-list --no-walk 5f95be55e3327417cea16f2cf418bad7fa91b705 # timeout=10
> Parsing POMs
> Established TCP socket on 44360
> maven3-agent.jar already up to date
> maven3-interceptor.jar already up to date
> maven3-interceptor-commons.jar already up to date
> [Middleware-Parent] $ /usr/lib/jvm/java-8-oracle//bin/java -cp /home/jenkins/maven3-agent.jar:/usr/share/maven/boot/plexus-classworlds-2.x.jar org.jvnet.hudson.maven3.agent.Maven3Main /usr/share/maven /home/jenkins/remoting-3.14.jar /home/jenkins/maven3-interceptor.jar /home/jenkins/maven3-interceptor-commons.jar 44360
> <===[JENKINS REMOTING CAPACITY]===>���channel started
> Executing Maven:  -B -f /home/jenkins/workspace/Middleware_Develop/Middleware-Parent/pom.xml clean install -rf karaf-feature -P apiDocBuilder,buildInstaller,karaf-assembly -DskipTests=true -Dproduct.name=QuareoMiddleware -Dbuild-info.buildNumber=Middleware_Develop-43 -Dhostname=docker-29a83953db3cf
> [INFO] Scanning for projects...
> [WARNING] 
> [WARNING] Some problems were encountered while building the effective model for org.codehaus.izpack:IzPack:pom:4.3.5
> [WARNING] 'build.plugins.plugin.version' for org.apache.maven.plugins:maven-install-plugin is missing. @ line 102, column 13
> [WARNING] 
> [WARNING] Some problems were encountered while building the effective model for com.te.middleware.installer.karaf:MiddlewareInstallerKarafAssembly:jar:3.0.2-SNAPSHOT
> [WARNING] 'build.plugins.plugin.(groupId:artifactId)' must be unique but found duplicate declaration of plugin org.apache.maven.plugins:maven-dependency-plugin @ line 71, column 21
> [WARNING] 
> [WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
> [WARNING] 
> [WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
> [WARNING] 
> [INFO] Initializing Artifactory Build-Info Recording
> [INFO] ------------------------------------------------------------------------
> [INFO] Reactor Build Order:
> [INFO] 
> [INFO] Quareo Middleware :: Karaf Feature
> [INFO] Quareo Middleware :: Karaf Runtime
> [INFO] IzPack Standalone Compiler
> [INFO] Quareo Middleware Installer Parent
> [INFO] Quareo Middleware Installer
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> [HUDSON] Collecting dependencies info
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> [HUDSON] Collecting dependencies info
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> [HUDSON] Collecting dependencies info
> [HUDSON] Collecting dependencies info
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> [HUDSON] Collecting dependencies info
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> Notifying upstream projects of job completion
> Join notifier requires a CauseAction
> ERROR: Build step failed with exception
> java.lang.NullPointerException: no workspace from node io.jenkins.docker.DockerTransientNode[docker-29a83953db3cf] which is computer DockerComputer{name=docker-29a83953db3cf, slave=io.jenkins.docker.DockerTransientNode[docker-29a83953db3cf]} and has channel null
> 	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:88)
> 	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:66)
> 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
> 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:744)
> 	at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.build(MavenModuleSetBuild.java:945)
> 	at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.doRun(MavenModuleSetBuild.java:896)
> 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:504)
> 	at hudson.model.Run.execute(Run.java:1724)
> 	at hudson.maven.MavenModuleSetBuild.run(MavenModuleSetBuild.java:543)
> 	at hudson.model.ResourceController.execute(ResourceController.java:97)
> 	at hudson.model.Executor.run(Executor.java:429)
> Build step 'Execute shell' marked build as failure
> ERROR: Failed to parse POMs
> java.io.IOException: Backing channel 'docker-29a83953db3cf' is disconnected.
> 	at hudson.remoting.RemoteInvocationHandler.channelOrFail(RemoteInvocationHandler.java:212)
> 	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:281)
> 	at com.sun.proxy.$Proxy98.isAlive(Unknown Source)
> 	at hudson.Launcher$RemoteLauncher$ProcImpl.isAlive(Launcher.java:1138)
> 	at hudson.maven.ProcessCache$MavenProcess.call(ProcessCache.java:166)
> 	at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.doRun(MavenModuleSetBuild.java:879)
> 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:504)
> 	at hudson.model.Run.execute(Run.java:1724)
> 	at hudson.maven.MavenModuleSetBuild.run(MavenModuleSetBuild.java:543)
> 	at hudson.model.ResourceController.execute(ResourceController.java:97)
> 	at hudson.model.Executor.run(Executor.java:429)
> Caused by: java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 1
> 	at io.jenkins.docker.client.DockerMultiplexedInputStream.readInternal(DockerMultiplexedInputStream.java:45)
> 	at io.jenkins.docker.client.DockerMultiplexedInputStream.read(DockerMultiplexedInputStream.java:32)
> 	at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:99)
> 	at hudson.remoting.ChunkedInputStream.read(ChunkedInputStream.java:46)
> 	at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:97)
> 	at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:39)
> 	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)
> 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)
> ERROR: Step ‘JIRA: Update relevant issues’ failed: no workspace for Middleware_Develop #43
> ERROR: Step ‘Publish Cobertura Coverage Report’ failed: no workspace for Middleware_Develop #43
> Notifying upstream projects of job completion
> Finished: FAILURE

The underlying Jenkins log error for the stack trace above is

> Feb 26, 2018 9:50:09 AM SEVERE hudson.remoting.SynchronousCommandTransport$ReaderThread run
> I/O error in channel Channel to Maven [/usr/lib/jvm/java-8-oracle//bin/java, -cp, /home/jenkins/maven3-agent.jar:/usr/share/maven/boot/plexus-classworlds-2.x.jar, org.jvnet.hudson.maven3.agent.Maven3Main, /usr/share/maven, /home/jenkins/remoting-3.14.jar, /home/jenkins/maven3-interceptor.jar, /home/jenkins/maven3-interceptor-commons.jar, 44360]
> java.io.IOException: Expected 8 bytes application/vnd.docker.raw-stream header, got 1
> 	at io.jenkins.docker.client.DockerMultiplexedInputStream.readInternal(DockerMultiplexedInputStream.java:45)
> 	at io.jenkins.docker.client.DockerMultiplexedInputStream.read(DockerMultiplexedInputStream.java:32)
> Caused: java.io.IOException
> 	at hudson.remoting.FastPipedInputStream.read(FastPipedInputStream.java:169)
> 	at java.io.BufferedInputStream.fill(Unknown Source)
> 	at java.io.BufferedInputStream.read1(Unknown Source)
> 	at java.io.BufferedInputStream.read(Unknown Source)
> 	at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:99)
> 	at hudson.remoting.ChunkedInputStream.read(ChunkedInputStream.java:46)
> 	at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:97)
> 	at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:39)
> 	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)
> 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)

[clouds.txt](https://github.com/jenkinsci/docker-plugin/files/1759108/clouds.txt)

[logs.zip](https://github.com/jenkinsci/docker-plugin/files/1759227/logs.zip)
This set of log files covers the docker inspect for the slave I'm building on, the docker logs (effectively empty), the Jenkins log, showing the underlying problem, and a trunated version of the Build console for the failed build.

If I can provide any information you think I've missed, please just respond and I'll try to provide it.

Many thanks.

Pete.

--- a/src/main/java/io/jenkins/docker/client/DockerMultiplexedInputStream.java
+++ b/src/main/java/io/jenkins/docker/client/DockerMultiplexedInputStream.java
@@ -38,12 +38,12 @@ public class DockerMultiplexedInputStream extends InputStream {
     private void readInternal() throws IOException {
         while (next == 0) {
             byte[] header = new byte[8];
-            int i = multiplexed.read(header);
-            if (i < 0) return; // EOF
-
-            if (i != 8)
-                throw new IOException("Expected 8 bytes application/vnd.docker.raw-stream header, got " + i);
-
+            int todo = 8;
+            while (todo > 0) {
+                int i = multiplexed.read(header, 8-todo, todo);
+                if (i < 0) return; // EOF
+                todo -= i;
+            }
             int size = ((header[4] & 0xff) << 24) + ((header[5] & 0xff) << 16) + ((header[6] & 0xff) << 8) + (header[7] & 0xff);
             switch (header[0]) {
                 case 1: // STDOUT
@@ -54,7 +54,7 @@ public class DockerMultiplexedInputStream extends InputStream {
                     byte[] payload = new byte[size];
                     int received = 0;
                     while (received < size) {
-                        i = multiplexed.read(payload, received, size - received);
+                        int i = multiplexed.read(payload, received, size - received);
                         received += i;
                     }
                     System.err.println(new String(payload, StandardCharsets.UTF_8));

--- a/src/main/java/io/jenkins/docker/client/DockerMultiplexedInputStream.java
+++ b/src/main/java/io/jenkins/docker/client/DockerMultiplexedInputStream.java
@@ -55,9 +55,10 @@ public class DockerMultiplexedInputStream extends InputStream {
                     int received = 0;
                     while (received < size) {
                         int i = multiplexed.read(payload, received, size - received);
+                        if (i < 0) break; // EOF
                         received += i;
                     }
-                    System.err.println(new String(payload, StandardCharsets.UTF_8));
+                    System.err.println(new String(payload, 0, received, StandardCharsets.UTF_8));
                     break;
                 default:
                     throw new IOException("Unexpected application/vnd.docker.raw-stream frame type " + header);

--- a/src/main/java/io/jenkins/docker/client/DockerMultiplexedInputStream.java
+++ b/src/main/java/io/jenkins/docker/client/DockerMultiplexedInputStream.java
@@ -4,6 +4,9 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 /**
  * De-multiplex an <code>application/vnd.docker.raw-stream</code> as described on
  * <a href="https://docs.docker.com/engine/api/v1.32/#operation/ContainerAttach">Docker API documentation</a>
@@ -14,6 +17,8 @@ public class DockerMultiplexedInputStream extends InputStream {
 
     private final InputStream multiplexed;
     int next;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DockerMultiplexedInputStream.class);
 
     public DockerMultiplexedInputStream(InputStream in) {
         multiplexed = in;
@@ -50,7 +55,7 @@ public class DockerMultiplexedInputStream extends InputStream {
                     next = size;
                     break;
                 case 2: // STDERR
-                    // not expected. Read payload and dump on stderr for diagnostic
+                    // not expected. Read payload and log it for diagnostic
                     byte[] payload = new byte[size];
                     int received = 0;
                     while (received < size) {
@@ -58,7 +63,8 @@ public class DockerMultiplexedInputStream extends InputStream {
                         if (i < 0) break; // EOF
                         received += i;
                     }
-                    System.err.println(new String(payload, 0, received, StandardCharsets.UTF_8));
+                    LOGGER.warn("Unexpected data on container stderr: {}",
+                                new String(payload, 0, received, StandardCharsets.UTF_8));
                     break;
                 default:
                     throw new IOException("Unexpected application/vnd.docker.raw-stream frame type " + header);

--- a/src/test/java/io/jenkins/docker/client/DockerMultiplexedInputStreamTest.java
+++ b/src/test/java/io/jenkins/docker/client/DockerMultiplexedInputStreamTest.java
@@ -1,0 +1,321 @@
+package io.jenkins.docker.client;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.io.ByteArrayOutputStream;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+
+import java.util.Arrays;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class DemuxTester implements AutoCloseable, Runnable {
+
+    // helper class for testing DockerMultiplexedInputStream
+    //
+    //                .--------------------------------.
+    //                |          this.feeder           |
+    //  ------------->|      (PipedOutputStream)       |
+    //  input         .--------------------------------.
+    //                |    this.feeder_input_stream    |
+    //                |       (PipedInputStream)       |
+    //                '--------------------------------'
+    //                                 |
+    //                                 v
+    //                .--------------------------------.
+    //                |          this.stream           |
+    //                | (DockerMultiplexedInputStream) |
+    //                '--------------------------------'
+    //                                 |
+    //                                 v
+    //                  .-----------------------------.
+    //                  |         this.thread         |
+    //                  |                             |
+    //                  | [reads from this.stream     |
+    //                  |  and writes into this.sink] |
+    //                  '-----------------------------'
+    //                                 |
+    //                                 v
+    //                .--------------------------------.
+    //                |           this.sink            |
+    //  <-------------|    (ByteArrayOutputStream)     |
+    //  expected      '--------------------------------'
+    //  output    
+    
+    PipedOutputStream feeder;
+    PipedInputStream  feeder_input_stream;
+    DockerMultiplexedInputStream stream;
+    ByteArrayOutputStream sink;
+    Thread thread;
+    boolean eof;
+    IOException exc;
+
+    public DemuxTester() throws IOException {
+        feeder = new PipedOutputStream();
+
+        feeder_input_stream = new PipedInputStream();
+        feeder_input_stream.connect(feeder);
+        
+        stream = new DockerMultiplexedInputStream(feeder_input_stream);
+        sink = new ByteArrayOutputStream();
+        eof = false;
+        exc = null;
+
+        thread = new Thread(this);
+        thread.start();
+    }
+
+    // DemuxTester thread function
+    //
+    // forwards this.stream into this.sink
+    @Override
+    public void run() {
+        try
+        {
+            byte[] buffer = new byte[64];
+            int count;
+            while(true) {
+                count = stream.read(buffer, 0, buffer.length);
+                if (count <= 0) {
+                    // Demux EOF
+                    eof = true;
+                    return;
+                }
+                sink.write(buffer, 0, count);
+                sink.flush();
+            }
+        }
+        catch (IOException e) { 
+            exc = e;
+            e.printStackTrace();
+        }
+    }
+
+    // return true if the DockerMultiplexedInputStream is at EOF
+    public boolean isEof() {
+        return eof;
+    }
+
+    // return the exception if set
+    public IOException exception() {
+        return exc;
+    }
+
+    @Override
+    public void close() {
+        try {
+            feeder.close(); 
+            thread.join(100);
+        }
+        catch (IOException | InterruptedException e) { 
+            e.printStackTrace();
+        }
+    }
+
+    // run an iteration of the test
+    public void iteration(byte[] input, byte[] expected_output) throws IOException, InterruptedException
+    {
+        iteration(input, expected_output, false);
+    }
+    public void iteration(byte[] input, byte[] expected_output, boolean send_eof) throws IOException, InterruptedException
+    {
+        //  1. feed the 'input' bytes into the channel
+        feeder.write(input);
+        feeder.flush();
+        if (send_eof) {
+            feeder.close();
+        }
+
+        //  2. wait until all bytes are processed
+        //     (with a 1-second timeout to prevent hanging)
+        for (int i=200 ; i!=0 ; i--) {
+            if (feeder_input_stream.available() == 0) {
+                break;
+            }
+            Thread.sleep(5);
+        }
+        Thread.sleep(5);
+
+        //  3. ensure the output equals 'expected_output'
+        Assert.assertArrayEquals(expected_output, sink.toByteArray());
+        sink.reset();
+    }
+}
+
+public class DockerMultiplexedInputStreamTest {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DockerMultiplexedInputStream.class);
+
+    @Test
+    public void testIncompleteFrames() throws Exception {
+        try (DemuxTester tester = new DemuxTester())
+        {
+            tester.iteration(new byte[]{
+                    // full frame (payload: 3 bytes)
+                    1,0,0,0, 0,0,0,3, 65,66,67,
+                    // incomplete frame (payload: 3/7 bytes)
+                    1,0,0,0, 0,0,0,7, 68,69,70,
+                },
+                new byte[]{65,66,67,68,69,70});
+            
+            tester.iteration(new byte[]{
+                    // end of previous frame (4 bytes)
+                    71,72,73,74,
+                    // incomplete header (6/8 bytes)
+                    1,0,0,0,0,0,
+                },
+                new byte[]{71,72,73,74});
+
+            tester.iteration(new byte[]{
+                // end of header (2 bytes) + payload (3 bytes)
+                0,3, 75,76,77
+                },
+                new byte[]{75,76,77});
+
+            Assert.assertFalse(tester.isEof());
+            Assert.assertTrue(tester.exception() == null);
+        }
+    }
+
+
+    void subtestLargeFrame(DemuxTester tester, int size) throws Exception
+    {
+        // feed the frame header
+        tester.iteration(new byte[]{1, 0, 0, 0,
+            (byte)(size >> 24),
+            (byte)(size >> 16),
+            (byte)(size >> 8),
+            (byte) size},
+            new byte[]{}
+        );
+
+        // buffer with dummy data
+        byte[] buf = new byte[200];
+        for (int i=0 ; i<buf.length ; i++) {
+            buf[i] = (byte)i;
+        }
+
+        while (size > 0) {
+            // feed the frame payload (in 200-bytes chunk)
+            if (size >= 200 ) {
+                tester.iteration(buf, buf);
+                size -= 200;
+            } else {
+                byte[] slice = Arrays.copyOfRange(buf, 0, size);
+                tester.iteration(slice, slice);
+                size = 0;
+            }
+        }
+    }
+
+    @Test
+    public void testLargeFrames() throws Exception {
+        try (DemuxTester tester = new DemuxTester())
+        {
+            subtestLargeFrame(tester, 24);
+            subtestLargeFrame(tester, 327);
+            subtestLargeFrame(tester, 76893);
+            subtestLargeFrame(tester, 7);
+
+            Assert.assertFalse(tester.isEof());
+            Assert.assertTrue(tester.exception() == null);
+        }
+    }
+
+    @Test
+    public void testStderr() throws Exception {
+        try (DemuxTester tester = new DemuxTester())
+        {
+            tester.iteration(new byte[]{
+                    // stdout frame (3 bytes)
+                    1,0,0,0, 0,0,0,3, 65, 66, 67,
+                    // stderr frame (4 bytes)
+                    2,0,0,0, 0,0,0,4, 68, 69, 70, 71,
+                    // stdout frame (5 bytes)
+                    1,0,0,0, 0,0,0,5, 72, 73, 74, 75, 76,
+                },
+                // output (8 bytes)
+                new byte[]{65,66,67,72,73,74,75,76});
+
+            Assert.assertFalse(tester.isEof());
+            Assert.assertTrue(tester.exception() == null);
+        }
+    }
+
+    @Test
+    public void testEof() throws Exception {
+        // EOF in header
+        try (DemuxTester tester = new DemuxTester())
+        {
+            tester.iteration(new byte[]{
+                    // stdout frame (3 bytes)
+                    1,0,0,0, 0,0,0,3, 65, 66, 67,
+                    // incomplete header
+                    1,0,0,0, 0,0,
+                },
+                new byte[]{65,66,67}, 
+                // send EOF
+                true);
+            Assert.assertTrue(tester.isEof());
+            Assert.assertTrue(tester.exception() == null);
+        }
+
+        // EOF in stdout
+        try (DemuxTester tester = new DemuxTester())
+        {
+            tester.iteration(new byte[]{
+                    // stdout frame (3 bytes)
+                    1,0,0,0, 0,0,0,3, 65, 66, 67,
+                    // incomplete stdout frame (2/3 bytes)
+                    1,0,0,0, 0,0,0,3, 68, 69,
+                },
+                new byte[]{65,66,67, 68, 69}, 
+                // send EOF
+                true);
+            Assert.assertTrue(tester.isEof());
+            Assert.assertTrue(tester.exception() == null);
+        }
+
+        // EOF in stderr
+        
+        // log a dummy line to ensure the logger is initialised
+        // (otherwise the test fails because of a race condition)
+        LOGGER.warn("dummy log"); 
+
+        try (DemuxTester tester = new DemuxTester())
+        {
+            tester.iteration(new byte[]{
+                    // stdout frame (3 bytes)
+                    1,0,0,0, 0,0,0,3, 65, 66, 67,
+                    // incomplete stderr frame (2/3 bytes)
+                    2,0,0,0, 0,0,0,3, 68, 69,
+                },
+                new byte[]{65,66,67}, 
+                // send EOF
+                true);
+            Assert.assertTrue(tester.isEof());
+            Assert.assertTrue(tester.exception() == null);
+        }
+    }
+
+    @Test
+    public void testUnknownFrameType() throws Exception {
+        // EOF in header
+        try (DemuxTester tester = new DemuxTester())
+        {
+            tester.iteration(new byte[]{
+                // stdout frame (3 bytes)
+                1,0,0,0, 0,0,0,3, 65, 66, 67,
+                    // unknown frame (type 4) 
+                    4,0,0,0, 0,0,0,2, 68, 69,
+            },
+            new byte[]{65,66,67}); 
+
+            Assert.assertFalse(tester.isEof());
+            Assert.assertTrue(tester.exception() != null);
+        }
+    }
+}

